diff --git a/recipes-devtools/sdcard-raw-tools/sdcard-raw-tools/create_sdcard_from_flashlayout.sh b/recipes-devtools/sdcard-raw-tools/sdcard-raw-tools/create_sdcard_from_flashlayout.sh
index a426c37..10521fc 100755
--- a/recipes-devtools/sdcard-raw-tools/sdcard-raw-tools/create_sdcard_from_flashlayout.sh
+++ b/recipes-devtools/sdcard-raw-tools/sdcard-raw-tools/create_sdcard_from_flashlayout.sh
@@ -38,7 +38,7 @@ SDCARD_TOKEN=mmc0
 DEFAULT_RAW_SIZE=1536
 
 # size of 768MB
-DEFAULT_ROOTFS_PARTITION_SIZE=768432
+DEFAULT_ROOTFS_PARTITION_SIZE=2097152
 # size of 1024MB
 #DEFAULT_ROOTFS_PARTITION_SIZE=1232896
 
